### DISCUSSION TITLE
Export all notes and tags in CSV or JSON format

### DISCUSSION THREAD

**Discussion by debritto** [ezra-bible-app/ezra-bible-app#476] [Ideas]
Hi folks,
Is it possible to create a option to export all notes and tags in a CSV or JSON format?

All that I need is to export all this data:
```
SELECT   "BibleBooks"."shortTitle" as BibleBook,  "VerseReferences"."chapter",  "VerseReferences"."verseNr",  "Tags"."title" as Tag,   "Notes"."text" as Note FROM   "BibleBooks" INNER JOIN "VerseReferences"  ON "BibleBooks"."id" = "VerseReferences"."bibleBookId" INNER JOIN "VerseTags"  ON "VerseReferences"."id" = "VerseTags"."verseReferenceId" INNER JOIN "Tags"  ON "VerseTags"."tagId" = "Tags"."id" INNER JOIN "Notes"  ON "VerseReferences"."id" = "Notes"."verseReferenceId" 
```

My plan is to use this information to create some charts and make some analysis.
I am looking forward to hearing from you.
Blessings in Jesus name.

**Comment by tobias-klein** [comment]
Thanks for raising this @debritto!

Here are some questions for clarification:

1) Since this is a feature request for rather advanced users, would it be ok to only support it via command line parameter and not offer a UI button / menu entry for this?
2) When you say "all notes and tags" you mean a complete dump of all user data for all Bible books, right?
3) We could support both CSV and JSON, but need to pick one to start with. How about starting with JSON?

I think it could be feasible to implement this in 1.4 (planned for December).

@zhuiks Do you see any additional questions that should be clarified regarding this request?

**Comment by debritto** [comment]
Hi @tobias-klein !

Thank you for consider my suggestion. I'm very happy. 

1. I agree with you that It's better not to use a UI button or menu entry, in order to keep the minimalist principle developed for Ezra Bible App (I belive this is a very good charactheristic). Yes, it turns to be a advanced user feature. My aim is to enable users to export their tags and notes in a format that could be handled in [CAQDAS](https://en.wikipedia.org/wiki/Computer-assisted_qualitative_data_analysis_software) softwares type. I personally developing one for me and my students as an Open Source app. What do you think of the possibility to put a Export all link at "Module and Application Info -> Application Info" window, below "Configuration path" info. When the user click a link, let us say "Export all tags and notes" the App open the file window so the user could choose the name and file path".
2. Yes, when I say "all notes and tags" I mean a complete dump of all user data for all Bible Books. Everything.
3. It will be perfect to begin with JSON. Great!

Will be awesome if this feature could be implemented at the end of the year. I'll start to work with my students at the beginning of 2022. :)

Just another thing. Do you think that is possible to export the content of the related Bible verses to (At least to Bible translations in Public Domain)?

**Comment by debritto** [comment]
Oops. I forgot to say that my students will use the "Export all" option and they aren't advanced users. In a matter of fact, I believe that most users that could use this feature will be plain users. ;)

**Comment by tobias-klein** [comment]
> Just another thing. Do you think that is possible to export the content of the related Bible verses to (At least to Bible translations in Public Domain)?

Yes, should be possible. We will simply show a disclaimer to the user based on the modules with Copyright. We are doing the same in case of #96.

**Comment by zhuiks** [comment]
Here are my thoughts:
1. Regarding UI, I think the most natural/minimalistic would be to add "Export All" under dropdown next to "Export to Word" button: 
![Screenshot from 2021-10-15 20-14-25](https://user-images.githubusercontent.com/3584119/137568136-5f6e1e9f-755c-45ea-8d21-dca0134aeaa2.png)
2. I think the CSV format is more universal than JSON for non-technical users. It can be open with any spreadsheets program.
3. I definitely agree with the "Export All" concept - one click to dump all user-generated data. I'm a bit more skeptical about including Bible verses content. It's not "user-generated". From my experience, end-users either care too much about copyright or don't care at all (with or without disclaimer dialog). However, I guess adding Bible text to dump for Public Domain or some other open licenses. As it is publicly available anyway.

I think it would be possible to implement for the 1.4 version.

**Comment by tobias-klein** [comment]
I like your proposal @zhuiks. And I agree that CSV is easier to handle for the average user. However, the question is what's the purpose of this feature. If the export shall be used for further computer-based data processing (that's what I understood from @debritto), then a format like JSON may be more suitable, because you can "imitate" multiple database tables via that format (like having a list of tags and then along with that a list of tag assignments that reference those tags). Also, if we agree on a certain JSON format then this could be discussed with other frontend's developers and we may be able to establish a common format for export (and potentially also import). Furthermore, if the purpose is rather technical then for me the question is whether this needs to be accessible via a standard menu option or can also be a bit hidden, because it's very likely not used very often ...

**Comment by tobias-klein** [comment]
I turned this into a discussion while we are aligning on the requirements. Still a bit fuzzy at this point.

**Comment by debritto** [comment]
Great proposal @zhuiks  
1. I believe that an "Export All" under dropdown nest to "Export to Word" button will be better. 
2. Yes, CSV format is more universal than JSON to end users, and It can be handled with ease in any kind of spreadsheet app. (Considering this perspective, CSV is better than JSON. )
3. Will be better to export all user data only. In this way we would avoid any kind of problem with copyrighted material. Besides this, the process will be faster.

**Comment by zhuiks** [comment]
> what's the purpose of this feature?

From our discussions here, I understand it's one of these two:
1. Technical exchange (export/import) between different software. Then, what you [said](https://github.com/ezra-bible-app/ezra-bible-app/discussions/476#discussioncomment-1486990), @tobias-klein make sense, and we should see if a certain JSON format would suit other  SWORD frontends developers. I also agree that in this case, this feature should be a little more hidden.
2. Dump of all user data for whatever reason they have (switch to other software, further processing, print out). Then, CSV format is preferred. In this case it can be under "second click" next to export button as I proposed.

I think, that for now the second option is better. Based on [initial comment](https://github.com/ezra-bible-app/ezra-bible-app/discussions/476#discussion-3629087) it can be the following CSV table:

| Bible Book | Chapter | Verse | Tags | Note |
| -------------- | ----------- | -------- | ------ | ------ |
| John | 3 | 16 | God's Love; Eternal life |  |
| Ephesians |  |  |   | Written: \_aprox. \*\*60-63\*\* A.D.\_ |
| Ephesians | 1 | 1 | Will of God; Faithfulness | \_Apostle\_ - messenger, a delegate, commissioner of Christ |

@debritto, is this something that will suit your needs with the students?

Please note, that:
1. There might be multiple tags per verse.
2. There might be a note for the whole book (e.g. Ephesians).
3.  We store notes internally in Markdown format. So the export/dump will be in this format. Right now entering notes is also in this format, but later we might replace it with a simple WYSIWYG editor.

**Reply by debritto** [reply]
Hi @zhuiks !

I stick with the second option. I mean, dump all user data so they can use in different purposes. Specifically for my students the second option is better because they are not advanced users. The table you gave as sample is perfect. The markdown notes and rows occasionally without tags or even note content is not a problem.
Warm regards from Brazil!
